packages/flutter/lib/src/widgets/restoration.dart

```diff
@@ -0,0 +1,1031 @@
+/// [RestorationMixin] before the owning [State] object is disposed by calling
+/// [RestorationMixin.unregisterFromRestoration]. This is uncommon, though, and
+/// will delete the information that the property contributed from the
+/// restoration data (meaning the value of the property will no longer be
+/// restored during a future state restoration).
+///
+/// See also:
+///
+///  * [RestorableValue], which is a [RestorableProperty] that makes the wrapped
+///    value accessible to the owning [State] object via a `value`
+///    getter and setter.
+///  * [RestorationMixin], to which a [RestorableProperty] must be registered.
+///  * [RestorationManager], which describes how state restoration works in
+///    Flutter.
+abstract class RestorableProperty<T> extends ChangeNotifier {
+  /// Called by the [RestorationMixin] with the `value` returned by either
+  /// [createDefaultValue] or [fromPrimitives] to set the value that this
+  /// property currently wraps.
+  ///
+  /// The [initWithValue] method may be called multiple times throughout the
+  /// life of the [RestorableProperty] whenever new restoration data has been
+  /// provided to the [RestorationMixin] the property is registered to. When
+  /// [initWithValue] is called, the property should forget its previous value
+  /// and re-initialize itself to the newly provided `value`.
+  void initWithValue(T value);
+
+  /// Called by the [RestorationMixin] if no restoration data is available to
+  /// restore the value of the property from to obtain the default value for the
+  /// property.
+  ///
+  /// The method returns the default value that the property should wrap if no
+  /// restoration data is available. After this method has been called,
+  /// [initWithValue] will be invoked with the returned value.
+  ///
+  /// The method may be called multiple times throughout the life of the
+  /// [RestorableProperty]. Whenever new restoration data has been provided to
+  /// the [RestorationMixin] the property is registered to, either this method
+  /// or [fromPrimitives] is called before [initWithValue] is invoked.
+  T createDefaultValue();
+
+  /// Called by the [RestorationMixin] to retrieve the information that this
+  /// property wants to store in the restoration data.
+  ///
+  /// The returned object must be serializable with the [StandardMessageCodec]
+  /// and if it includes any collections, those should not be modified after
+  /// they have been returned by this method.
+  ///
+  /// The information returned by this method may be handed back to the property
+  /// in a call to [fromPrimitives] at a later point in time (possibly after the
+  /// application restarted) to restore the value that the property is currently
+  /// wrapping.
+  ///
+  /// When the value returned by this method changes, the property must call
+  /// [notifyListeners]. The [RestorationMixin] will invoke this method whenever
+  /// the property's listeners are notified.
+  Object toPrimitives();
```

Comment:
is it expected that `x.fromPrimitives(x.toPrimitives()) == x.fromPrimitives(x.toPrimitives())` ?

Reply:
Yes.